### PR TITLE
style: support calc() in color functions.

### PR DESCRIPTION
**Pull request by emilio**
Depends on #19456 and https://github.com/servo/rust-cssparser/pull/207.

Fixes https://bugzilla.mozilla.org/show_bug.cgi?id=984021

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19457)

### CODE REVIEW

File: ports/geckolib/glue.rs [hunk 1]
```diff
@@ -4552,7 +4552,16 @@ pub unsafe extern "C" fn Servo_SelectorList_Drop(list: RawServoSelectorListOwned
 fn parse_color(value: &str) -> Result<specified::Color, ()> {
```
Comment (SimonSapin) [inline ports/geckolib/glue.rs]: Is every use of this function intended to support full CSS syntax, including `calc`? I traced one use back to `aBGColor` in `DocumentRendererChild::RenderDocument`. Is that `<body bgcolor>`? https://html.spec.whatwg.org/multipage/#the-page links to https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#rules-for-parsing-a-legacy-colour-value

Comment (emilio) [reply ports/geckolib/glue.rs]: It is not, that's handled from `nsAttrValue::ParseColor` (https://searchfox.org/mozilla-central/rev/0b613c3887789f7786cd3131dfe9648398f4a6ac/dom/html/HTMLBodyElement.cpp#58).

That seems to be the default background color to render in a child process if there's no content.

The only web-exposed use of this is `CanvasGradient::AddColorStop`, and the spec for that is:

  https://html.spec.whatwg.org/multipage/#dom-canvasgradient-addcolorstop

Which says to parse as a CSS `<color>` value, that is, with `calc()`.

**Comment by emilio** [comment]
Sigh, I expected tests to be there already... I'll write some.

**Comment by SimonSapin** [comment]
r+ with one last change

---

Reviewed 14 of 14 files at r1, 1 of 1 files at r2, 1 of 4 files at r3, 7 of 7 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/style/values/specified/color.rs, line 162 at r6](https://reviewable.io:443/reviews/servo/servo/19457#-L-WJp4OZPvPoZCDuGZ_:-L-WJp4OZPvPoZCDuGZa:bb7zzmg) ([raw file](https://github.com/servo/servo/blob/3cd75d11a1ad04d6045ee96aec24b83db497fb6c/components/style/values/specified/color.rs#L162)):*
> ```Rust
>                     CSSParserColor::RGBA(rgba) => Color::Numeric {
>                         parsed: rgba,
>                         authored: authored.map(|s| s.to_lowercase().into_boxed_str()),
> ```

This probably doesn’t make a difference in behavior since all valid color keywords are ASCII, but this should be `to_ascii_lowercase`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/19457#-:-L-WKcLsbqTvOFZuGWu-:b-i23ypk)*

**Pull request merged**